src/support/small_set.h

```diff
@@ -34,10 +34,78 @@
 
 namespace wasm {
 
-template<typename T, size_t N, typename FlexibleSet> class SmallSetBase {
+template<typename T, size_t N> struct FixedStorageBase {
+  size_t used = 0;
+  std::array<T, N> storage;
+};
+
+template<typename T, size_t N>
+struct UnorderedFixedStorage : public FixedStorageBase<T, N> {
+  void insert(const T& x) {
+    assert(this->used < N);
+    this->storage[this->used++] = x;
+  }
+
+  void erase(const T& x) {
+    for (size_t i = 0; i < this->used; i++) {
+      if (this->storage[i] == x) {
+        // We found the item; erase it by moving the final item to replace it
+        // and truncating the size.
+        this->used--;
+        this->storage[i] = this->storage[this->used];
+        return;
+      }
+    }
+  }
+};
+
+template<typename T, size_t N>
+struct OrderedFixedStorage : public FixedStorageBase<T, N> {
+  void insert(const T& x) {
+    assert(this->used < N);
+
+    if (this->used == 0) {
+      this->storage[0] = x;
+      this->used = 1;
+      return;
+    }
+
+    // Find the insertion point |i| where x should be placed.
+    size_t i = 0;
+    while (i < this->used && this->storage[i] <= x) {
+      i++;
+    }
+    // |i| is now the location where x should be placed.
+
+    if (i != this->used) {
+      // Push things forward to make room for x.
+      for (size_t j = this->used; j >= i + 1; j--) {
+        this->storage[j] = this->storage[j - 1];
+      }
+    }
```

Comment:
Hmm, [docs](https://en.cppreference.com/w/cpp/algorithm/move_backward) say

> The behavior is undefined if d_last is within (first, last]. [std::move](https://en.cppreference.com/w/cpp/algorithm/move) must be used instead of std::move_backward in that case.

Even with move though I'm not sure it's nicer?
```diff
-      for (size_t j = this->used; j >= i + 1; j--) {
-        this->storage[j] = this->storage[j - 1];
-      }
+      std::move(this->storage.begin() + i,
+                this->storage.begin() + this->used,
+                this->storage.begin() + i + 1);
```

Reply:
We're pushing things back in the array here, so I think this is the correct up. `d_last` is one after `last`, so it is not within `(first, last]`.

Reply:
Hmm, I guess that's right. However, the test suite errors when I use `move_backwards`..! But it works with `move`.

I haven't looked into that in depth, but again, I'm just not sure replacing obvious low-level code with higher-level APIs that you need to read the spec of is worthwhile...

Reply:
Wow, what a pain.